sbom/create_product_sbom.py

```
@@ -1,159 +1,215 @@
+    component_relationships = get_component_relationships(product_elem_id, component_packages)
 
-    return sbom
+    return Document(
+        creation_info=creation_info,
+        packages=[product_package, *component_packages],
+        relationships=[product_relationship, *component_relationships],
+    )
 
 
-def main():
+def main() -> None:
+    """
+    Script entrypoint.
+    """
     parser = argparse.ArgumentParser(
-        prog="create-product-sbom", description="Create product-level SBOM from releaseNotes."
+        prog="create-product-sbom",
+        description="Create product-level SBOM from merged data file"
+        "and mapped snapshot spec.",
+    )
+    parser.add_argument(
+        "--data-path",
+        required=True,
+        type=Path,
+        help="Path to the merged data file in JSON format.",
     )
-
     parser.add_argument(
-        "--data-path", required=True, type=str, help="Path to the input data in JSON format."
+        "--snapshot-path",
+        required=True,
+        type=Path,
+        help="Path to the input data in JSON format.",
     )
     parser.add_argument(
         "--output-path",
         required=True,
-        type=str,
+        type=Path,
         help="Path to save the output SBOM in JSON format.",
     )
 
     args = parser.parse_args()
+    setup_sbom_logger()
+
+    try:
+        snapshot = asyncio.run(sbomlib.make_snapshot(args.snapshot_path))
+        with open(args.data_path, "r", encoding="utf-8") as fp:
+            raw_json = fp.read()
+            release_notes = ReleaseData.model_validate_json(raw_json).release_notes
+
+            sbom = create_sbom(release_notes, snapshot)
 
-    sbom = create_sbom(args.data_path)
-    with open(args.output_path, "w") as fp:
-        json.dump(sbom, fp)
+        write_file(document=sbom, file_name=str(args.output_path), validate=True)
+    except Exception:  # pylint: disable=broad-except
+        logger.exception("Creation of the product-level SBOM failed.")
```

That seems very broad. Is that a good idea? If we want to allow the script to fail, wouldn't it be better to handle it in the bash script calling it? Or if we go this way, shouldn't we at least print the exception for debugging purposes? With the current code you won't even know what happened.

---

The [exception()](https://docs.python.org/3/library/logging.html#logging.Logger.exception) method captures the raised exception message and logs it automatically, so we will always see the traceback.

---

Although, there is a missing `raise` statement here :+1:

---

Ah, ok. I assumed you intentionally wanted to swallow the exception.

> The [exception()](https://docs.python.org/3/library/logging.html#logging.Logger.exception) method captures the raised exception message and logs it automatically, so we will always see the traceback.

Good point.

---

Fixed.